src/pal/src/map/map.cpp

```diff
@@ -2193,96 +2272,170 @@ MAPmmapAndRecord(
+            palError = ERROR_INVALID_NAME;
+            goto done;
+        }
+
+        fd = InternalOpen(peName, O_RDONLY);
+
+        if (fd == -1)
+        {
+            palError = ERROR_INVALID_NAME;
+            goto done;
+        }
+    }
+    else
+    {
+        //Step 0: Verify values, find internal pal data structures.
+        if (INVALID_HANDLE_VALUE == hFile)
+        {
+            ERROR_(LOADER)( "Invalid file handle\n" );
+            palError = ERROR_INVALID_HANDLE;
+            goto done;
+        }
+
+        palError = g_pObjectManager->ReferenceObjectByHandle(
+                pThread,
+                hFile,
+                &aotFile,
+                GENERIC_READ,
+                &pFileObject
+                );
+        if (NO_ERROR != palError)
+        {
+            ERROR_(LOADER)( "ReferenceObjectByHandle failed\n" );
+            goto done;
+        }
+
+        palError = pFileObject->GetProcessLocalData(
+                pThread,
+                ReadLock,
+                &pLocalDataLock,
+                reinterpret_cast<void**>(&pLocalData)
+                );
+        if (NO_ERROR != palError)
+        {
+            ERROR_(LOADER)( "GetProcessLocalData failed\n" );
+            goto done;
+        }
+
+        fd = pLocalData->unix_fd;
     }
 
-    int fd;
-    fd = pLocalData->unix_fd;
     //Step 1: Read the PE headers and reserve enough space for the whole image somewhere.
     IMAGE_DOS_HEADER dosHeader;
     IMAGE_NT_HEADERS ntHeader;
     errno = 0;
     if (0 != lseek(fd, 0, SEEK_SET))
     {
-        palError = FILEGetLastErrorFromErrno();
-        ERROR_(LOADER)( "lseek failed\n" );
+        if (!doPreload)
+        {
+            palError = FILEGetLastErrorFromErrno();
+            ERROR_(LOADER)( "lseek failed\n" );
+        }
+        else
+        {
+            palError = ERROR_INVALID_NAME;
+        }
         goto done;
     }
     if (sizeof(dosHeader) != read(fd, &dosHeader, sizeof(dosHeader)))
     {
-        palError = FILEGetLastErrorFromErrno();
-        ERROR_(LOADER)( "reading dos header failed\n" );
+        if (!doPreload)
+        {
+            palError = FILEGetLastErrorFromErrno();
+            ERROR_(LOADER)( "reading dos header failed\n" );
+        }
+        else
+        {
+            palError = ERROR_INVALID_NAME;
+        }
         goto done;
     }
     if (dosHeader.e_lfanew != lseek(fd, dosHeader.e_lfanew, SEEK_SET))
     {
-        palError = FILEGetLastErrorFromErrno();
+        if (!doPreload)
+        {
+            palError = FILEGetLastErrorFromErrno();
+        }
+        else
+        {
+            palError = ERROR_INVALID_NAME;
+        }
         goto done;
     }
     if (sizeof(ntHeader) != read(fd, &ntHeader, sizeof(ntHeader)))
     {
-        palError = FILEGetLastErrorFromErrno();
+        if (!doPreload)
+        {
+            palError = FILEGetLastErrorFromErrno();
+        }
+        else
+        {
+            palError = ERROR_INVALID_NAME;
```

Comment:
This case seems to be unrelated to a name or path. What do you think about using a more general error code here?